impl/src/main/java/io/kroxylicious/testing/kafka/common/KafkaTopology.java

```diff
@@ -0,0 +1,237 @@
+        this.driver = driver;
+        this.config = config;
+        if (!config.isKraftMode()) {
+            this.zookeeperConfig = new ZookeeperConfig(driver);
+            this.zookeeper = driver.createZookeeper(zookeeperConfig);
+        }
+        else {
+            this.zookeeperConfig = null;
+            this.zookeeper = null;
+        }
+        this.nodes = new HashMap<>();
+        this.nodeConfigurations = new HashMap<>();
+    }
+
+    boolean isKraftMode() {
+        return this.config.isKraftMode();
+    }
+
+    public static KafkaTopology create(KafkaDriver driver, KafkaClusterConfig config) {
+        KafkaTopology kafkaTopology = new KafkaTopology(driver, config);
+        Map<Integer, KafkaNodeConfiguration> configs = generateConfigurations(driver, config, kafkaTopology);
+        kafkaTopology.nodeConfigurations.putAll(configs);
+        // createNode requires the topology to have all nodes configurations installed first, so it can ask for all
+        // controller node details.
+        Map<Integer, KafkaNode> nodes = configs.values().stream()
+                .collect(Collectors.toMap(KafkaNodeConfiguration::nodeId, driver::createNode));
+        kafkaTopology.addNodes(nodes);
+        return kafkaTopology;
+    }
+
+    @NotNull
+    private static Map<Integer, KafkaNodeConfiguration> generateConfigurations(KafkaDriver driver, KafkaClusterConfig config, KafkaTopology kafkaTopology) {
+        int nodesToCreate = Math.max(config.getBrokersNum(), config.getKraftControllers());
+        Stream<KafkaNodeConfiguration> nodeConfigurationStream = IntStream.range(0, nodesToCreate)
+                .mapToObj(value -> new IdAndRoles(roles(config, value), value))
+                // note that generation of a configuration may allocate resources like ports to it
+                .map(idAndRoles -> getNodeConfiguration(kafkaTopology, config, idAndRoles, driver));
+        return nodeConfigurationStream.collect(Collectors.toMap(KafkaNodeConfiguration::nodeId, configuration -> configuration));
+    }
+
+    private void addNodes(Map<Integer, KafkaNode> nodes) {
+        this.nodes.putAll(nodes);
+    }
+
+    public String getQuorumVoters() {
+        String collect = nodeDescriptions().stream().map(KafkaNodeConfiguration::controllerQuorumAddress)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(Collectors.joining(","));
+        if (collect.isBlank()) {
+            throw new IllegalStateException("quorum voters empty! we should have some controllers in the cluster");
+        }
+        return collect;
+    }
+
+    public void close() {
+        if (this.zookeeper != null) {
+            this.zookeeper.close();
+        }
+    }
+
+    public void startZookeeperIfRequired() {
+        if (zookeeper != null) {
+            zookeeper.start();
+        }
+    }
+
+    public record IdAndRoles(Set<Role> roles, int nodeId) {
```

Comment:
Calling it `IdAndRoles` and then having the arguments as `Roles, id` really messes with me...